typing/value_rec_check.ml

```diff
@@ -189,28 +208,130 @@ let classify_expression : Typedtree.expression -> sd =
         in
-        if Array.for_all is_constant_field fields then Constant else Static
+        if Array.for_all is_constant_field fields
+        then Constant
+        else begin
+          match representation with
+          | Record_regular | Record_inlined _ ->
+              Static (Regular_block (Array.length fields))
+          | Record_float ->
+              Static (Float_record (Array.length fields))
+          | Record_extension _ ->
+              Static (Regular_block (1 + (Array.length fields)))
+          | Record_unboxed _ -> assert false (* handled above *)
+        end
     | Texp_tuple exprs ->
-        if List.for_all is_constant exprs then Constant else Static
+        if List.for_all is_constant exprs
+        then Constant
+        else Static (Regular_block (List.length exprs))
 
     | Texp_apply ({exp_desc = Texp_ident (_, _, vd)}, _)
       when is_ref vd ->
-        Static
+        Static (Regular_block 1)
     | Texp_apply (_,args)
       when List.exists is_abstracted_arg args ->
-        Static
+        Static Function
     | Texp_apply _ ->
         Not_recursive
 
-    | Texp_new _
-    | Texp_instvar _
-    | Texp_array _
-    | Texp_pack _
-    | Texp_object _
-    | Texp_function _
-    | Texp_lazy _
+    | Texp_array fields ->
+        begin match Typeopt.array_kind e with
+        | Lambda.Pfloatarray ->
+          Static (Float_record (List.length fields))
+        | Lambda.Pgenarray | Lambda.Paddrarray | Lambda.Pintarray ->
+          Static (Regular_block (List.length fields))
+        end
+    | Texp_pack mexp ->
+        let size_of_structure (s : Typedtree.structure) =
+          let rec size_of_structure_items acc items =
+            match (items : Typedtree.structure_item list) with
+            | [] -> acc
+            | item :: rem ->
+              let acc =
+                match item.str_desc with
+                | Tstr_value (_rec_flag, pat_expr_list) ->
+                  acc + List.length (let_bound_idents pat_expr_list)
+                | Tstr_typext tyext ->
+                  acc + List.length tyext.tyext_constructors
+                | Tstr_exception _ext ->
+                  acc + 1
+                | Tstr_module ({mb_presence=Mp_present} as mb) ->
+                  begin match mb.mb_id with
+                  | None -> acc
+                  | Some _ -> acc + 1
+                  end
+                | Tstr_recmodule bindings ->
+                  let ids = (List.filter_map (fun mb -> mb.mb_id) bindings) in
+                  acc + List.length ids
+                | Tstr_class cl_list ->
+                  acc + List.length cl_list
+                | Tstr_include incl ->
+                  let ids = bound_value_identifiers incl.incl_type in
+                  acc + List.length ids
+                | Tstr_open od ->
+                  let ids = bound_value_identifiers od.open_bound_items in
+                  acc + List.length ids
```

Comment:
This case looks wrong from a distance, does `open` add any item to the current structure?

Reply:
See #12785.
Summary: `open` sometimes does add items to the current structure, but in that case they must be removed from the signature and a coercion is generated. Since for the size computation, we never look at structures under a coercion, in practice `ids` should always be empty when we reach this case.
I could add an assertion (and a comment to go with that), or leave it like this so that it obviously mirrors `Translmod`.

Reply:
I don't agree that the logic here mirrors the logic in translmod. The logic here is trying to compute the size of the *resulting* structure, while the code in translmod builds an *ambient environment* as a series of let-bindings, that is not related to the size of the output structure. The two cases here for `open` and `include` are arguably wrong: the case for `open` is clearly wrong¹, and I am doubtful about the case for `include` in cases such as `include struct (open struct let x = 1 end) end` where we include hidden items.

On the other hand, the type-checker computed a module signature that is stored with the coercion, and contains all the information we need to know the size of the resulting structure.

So I would prefer to see the following code:
- a `size_of_signature` function
- instead of computing of `Texp_pack mexp` from `mexp.mod_desc`, we compute it from `mexp.mod_type`.

Reply:
>     * a `size_of_signature` function

That looks reasonable. I'll see how easy it is, but hopefully it's something as simple as `List.length (bound_value_identifiers sg)`. I expect that it will give the same result as my computation though.

>     * instead of computing of `Texp_pack mexp` from `mexp.mod_desc`, we compute it from `mexp.mod_type`.

I decided against that because I didn't want to authorize more expressions than necessary (i.e. `let rec m = (module F(X))` should be forbidden in my opinion, even if we have an explicit signature). But I can still use `size_of_signature` on `Tmod_structure` blocks that don't already have a coercion.